storage/innobase/log/log0recv.cc

```diff
@@ -4855,19 +4873,21 @@ byte *recv_dblwr_t::find_page(const page_id_t page_id,
     }
 
     const lsn_t lsn= mach_read_from_8(page + FIL_PAGE_LSN);
-    if (lsn <= max_lsn ||
-        !validate_page(page_id, page, space, tmp_buf))
+    if (lsn > min_lsn ||
+        lsn < recv_sys.parse_start_lsn ||
+        !validate_page(page_id, page, space, tmp_buf, true))
     {
       /* Mark processed for subsequent iterations in buf_dblwr_t::recover() */
-      memset(page + FIL_PAGE_LSN, 0, 8);
+      memset_aligned<8>(page + FIL_PAGE_LSN, 0, 8);
       continue;
```

Comment:
When we were discussing this on Friday, I had missed the fact that recovery invokes `log_sys.set_lsn(recv_sys.recovered_lsn)` before it is reading any data pages. Therefore, we can actually rely on `log_sys.get_lsn()` being up to date when validating any `FIL_PAGE_LSN`.

This may indeed be fixing an anomaly where a too new page could be recovered from the doublewrite buffer. I am not sure if that is a possible scenario, because `buf_dblwr_t::recover()` will only attempt to recover pages that were corrupted in the final location. Writes to the final location should have been preceded by a durable write to the log file, which in turn would make the doublewrite copy safe for recovery.

Reply:
3000bbd35d4955b2cfc0d7dfb2221234d894e126 revises the doublewrite buffer recovery in such a way that before `buf_dbwlr.recover()` is invoked, no data pages will be written. That function is responsible for recovering all pages from the doublewrite buffer. By the time it is invoked, the final recovered LSN will be known and assigned to `log_sys.lsn`.